mc_rtc_rviz_panel/src/Schema.cpp

```diff
@@ -0,0 +1,319 @@
+#include "Schema.h"
+
+#include "SchemaArrayInput.h"
+
+#include <mc_rtc/logging.h>
+
+#include <boost/filesystem.hpp>
+namespace bfs = boost::filesystem;
+
+namespace mc_rtc_rviz
+{
+
+Schema::Schema(const std::string & file) : Schema(mc_rtc::Configuration{file}, file) {}
+
+Schema::Schema(const std::vector<Schema> & schemas)
+{
+  is_tuple_ = true;
+  create_form = [schemas](QWidget * parent, const mc_rtc::Configuration & data) {
+    std::vector<FormElement *> elems;
+    for(const auto & s : schemas)
+    {
+      for(auto & el : s.create_form(parent, data))
+      {
+        elems.push_back(el);
+      }
+    }
+    return elems;
+  };
+}
+
+Schema::Schema(const mc_rtc::Configuration & s, const std::string & source, const std::string & title, bool required_in)
+{
+  title_ = s("title", title.size() ? title : "No title property in " + source);
+  auto required = s("required", std::vector<std::string>{});
+  auto is_required = [&required](const std::string & label) {
+    return std::find(required.begin(), required.end(), label) != required.end();
+  };
+  /** Resolve a $ref entry into a Schema */
+  auto resolve_ref = [](const std::string & source, const std::string & ref_) -> const Schema & {
+    auto ref = ref_.substr(3); // remove leading /..
```

Comment:
Out of curiosity, why do we have that leading `/..`, is that a requirement of JSON schema syntax? I've always found it odd in the schemas to have to go back one more level of hierarchy than the actual relative path.

Reply:
I think this was only due to a glitch in the JS library I was using a while back to display the schemas as it wouldn't resolve the reference correctly otherwise (basically acting as if the file name was a directory in the URI) and unfortunately it sticked...